core/matrix/sellp.hpp

```diff
@@ -0,0 +1,324 @@
+ * @tparam IndexType  precision of matrix indexes
+ *
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class Sellp : public EnableLinOp<Sellp<ValueType, IndexType>>,
+              public EnableCreateMethod<Sellp<ValueType, IndexType>>,
+              public ConvertibleTo<Dense<ValueType>>,
+              public ReadableFromMatrixData<ValueType, IndexType>,
+              public WritableToMatrixData<ValueType, IndexType> {
+    friend class EnableCreateMethod<Sellp>;
+    friend class EnablePolymorphicObject<Sellp, LinOp>;
+    friend class Dense<ValueType>;
+
+public:
+    using EnableLinOp<Sellp>::convert_to;
+    using EnableLinOp<Sellp>::move_to;
+
+    using value_type = ValueType;
+    using index_type = IndexType;
+    using mat_data = matrix_data<ValueType, IndexType>;
+
+    void convert_to(Dense<ValueType> *other) const override;
+
+    void move_to(Dense<ValueType> *other) override;
+
+    void read(const mat_data &data) override;
+
+    void write(mat_data &data) const override;
+
+    /**
+     * Returns the values of the matrix.
+     *
+     * @return the values of the matrix.
+     */
+    value_type *get_values() noexcept { return values_.get_data(); }
+
+    /**
+     * @copydoc Sellp::get_values()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const value_type *get_const_values() const noexcept
+    {
+        return values_.get_const_data();
+    }
+
+    /**
+     * Returns the column indexes of the matrix.
+     *
+     * @return the column indexes of the matrix.
+     */
+    index_type *get_col_idxs() noexcept { return col_idxs_.get_data(); }
+
+    /**
+     * @copydoc Sellp::get_col_idxs()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const index_type *get_const_col_idxs() const noexcept
+    {
+        return col_idxs_.get_const_data();
+    }
+
+    /**
+     * Returns the lengths(columns) of slices.
+     *
+     * @return the lengths(columns) of slices.
+     */
+    size_type *get_slice_lens() noexcept { return slice_lens_.get_data(); }
+
+    /**
+     * @copydoc Sellp::get_slice_lens()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const size_type *get_const_slice_lens() const noexcept
+    {
+        return slice_lens_.get_const_data();
+    }
+
+    /**
+     * Returns the offsets of slices.
+     *
+     * @return the offsets of slices.
+     */
+    size_type *get_slice_sets() noexcept { return slice_sets_.get_data(); }
```

Comment:
THis gives like a row-pointer to the beginning of each slice?

Reply:
@hartwiganzt 
Yes, it is like the column index if we append all slice by column.
The purpose is to save some computation since this information will be repeatedly used.

Reply:
column index? I thought it is like a row-pointer? Can you please give me a small example? Let's assume I have the sellp matrix (column indices):
/ 4 5 6 9
| 4 5 * *
| 1 2 5
| 2 3 9
| 3
\ 5

what would this function return?

Reply:
This is confusing me even more. The slice size is 2, I agree. the slice_legths would return 3 2 1/ But slice_sets? Would't that be the offset? Like, [ 0 8 14 16 ] ?